hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/storefiletracker/ModifyStoreFileTrackerProcedure.java

```diff
@@ -0,0 +1,269 @@
+ */
+package org.apache.hadoop.hbase.regionserver.storefiletracker;
+
+import java.io.IOException;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.DoNotRetryIOException;
+import org.apache.hadoop.hbase.HBaseIOException;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.TableNotEnabledException;
+import org.apache.hadoop.hbase.TableNotFoundException;
+import org.apache.hadoop.hbase.client.TableDescriptor;
+import org.apache.hadoop.hbase.master.procedure.AbstractStateMachineTableProcedure;
+import org.apache.hadoop.hbase.master.procedure.MasterProcedureEnv;
+import org.apache.hadoop.hbase.master.procedure.ModifyTableProcedure;
+import org.apache.hadoop.hbase.procedure2.ProcedureStateSerializer;
+import org.apache.hadoop.hbase.procedure2.ProcedureSuspendedException;
+import org.apache.hadoop.hbase.procedure2.ProcedureYieldException;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.hadoop.hbase.shaded.protobuf.ProtobufUtil;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.ModifyStoreFileTrackerState;
+import org.apache.hadoop.hbase.shaded.protobuf.generated.MasterProcedureProtos.ModifyStoreFileTrackerStateData;
+
+/**
+ * This procedure is used to change the store file tracker implementation.
+ * <p/>
+ * Typically we need to schedule two {@link ModifyTableProcedure} (or three if the table is already
+ * in {@code MIGRATION} but the {@code dstSFT} is not what we expected) to do this, so we introduce
+ * this procedure to simplify the work of our users.
+ */
+@InterfaceAudience.Private
+public abstract class ModifyStoreFileTrackerProcedure
+  extends AbstractStateMachineTableProcedure<ModifyStoreFileTrackerState> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ModifyStoreFileTrackerProcedure.class);
+
+  private TableName tableName;
+
+  private String dstSFT;
+
+  protected ModifyStoreFileTrackerProcedure() {
+  }
+
+  protected ModifyStoreFileTrackerProcedure(MasterProcedureEnv env, TableName tableName,
+    String dstSFT) throws HBaseIOException {
+    super(env);
+    checkDstSFT(dstSFT);
+    this.tableName = tableName;
+    this.dstSFT = dstSFT;
+    preflightChecks(env, true);
+  }
+
+  private void checkDstSFT(String dstSFT) throws DoNotRetryIOException {
+    if (MigrationStoreFileTracker.class
+      .isAssignableFrom(StoreFileTrackerFactory.getTrackerClass(dstSFT))) {
+      throw new DoNotRetryIOException("Do not need to transfer to " + dstSFT);
+    }
+  }
+
+  @Override
+  public TableName getTableName() {
+    return tableName;
+  }
+
+  @Override
+  public TableOperationType getTableOperationType() {
+    return TableOperationType.EDIT;
+  }
+
+  private enum StoreFileTrackerState {
+    NEED_RESTORE, NEED_MIGRATION, NEED_FINISH_MIGRATION, ALREADY_FINISHED
```

Comment:
`NEED_RESTORE` was a bit confusing to me and I had to read to understand what it meant. What about:

* `NEED_RESTORE` -> `NEED_SET_MIGRATION`
* `NEED_MIGRATION` -> `NEED_START_MIGRATION`
* `NEED_FINISH_MIGRATION` (perfect)

If you like this, should also change the constant up in proto.

Reply:
Finally I name them as

- `NEED_FINISH_PREVIOUS_MIGRATION_FIRST`
- `NEED_START_MIGRATION`
- `NEED_FINISH_MIGRATION`
- `ALREADY_FINISHED`

And on the procedure state, I name them as

- `MODIFY_STORE_FILE_TRACKER_FINISH_PREVIOUS_MIGRATION`
- `MODIFY_STORE_FILE_TRACKER_START_MIGRATION`
- `MODIFY_STORE_FILE_TRACKER_FINISH_MIGRATION`

Is this clear enough for you?
Thanks.